File: ruby/test/clock_test.rb
# frozen_string_literal: true

require_relative "helper"
require_relative "test_pki"

# The clock seam, and its boundaries. The injected clock is read in exactly
# one place: the endpoint's request_date triple. The JWS and receipt verifiers
# take none, and no payload is rejected for its age (PLAN.md D5). Everything
# else here exists to prove the clock reaches nowhere near a
# certificate-validity verdict.
class ClockTest < Minitest::Test
  APRV = ApplePurchaseReceiptVerifier

  SIGNED_AT = 1_722_945_600_000 # 2024-08-06T12:00:00Z

  def setup
    @pki = TestPki.jws_pki
    @jws = TestPki.sign_jws(@pki, TestPki.default_claims)
  end

  def verifier(**overrides)
    APRV::JwsVerifier.new(trusted_roots: [@pki.root],
                          bundle_id: "com.example.app",
                          accepted_environments: [APRV::Environment::SANDBOX], **overrides)
  end

  def test_a_payload_is_never_rejected_for_its_age
    assert_equal SIGNED_AT, verifier.verify_transaction(@jws).signed_date
  end

  def test_the_jws_verifier_takes_no_clock_and_no_max_age
    assert_raises(ArgumentError) { verifier(clock: -> { Time.utc(2020, 1, 1) }) }
    assert_raises(ArgumentError) { verifier(max_signed_age_seconds: 60) }
  end

  def test_a_dateless_payload_under_an_expired_chain_is_judged_at_now
    expired = TestPki.jws_pki(not_before: Time.utc(2019, 1, 1), not_after: Time.utc(2019, 12, 31))
    jws = TestPki.sign_jws(expired, TestPki.default_claims.except("signedDate"))
    subject = APRV::JwsVerifier.new(
      trusted_roots: [expired.root], bundle_id: "com.example.app",
      accepted_environments: [APRV::Environment::SANDBOX]
    )
    error = assert_raises(APRV::VerificationError) { subject.verify_transaction(jws) }
    assert_equal :INVALID_CHAIN, error.reason
  end

  def test_a_dateless_payload_under_a_live_chain_verifies
    jws = TestPki.sign_jws(@pki, TestPki.default_claims.except("signedDate"))
    assert_nil verifier.verify_transaction(jws).signed_date
  end

  def test_an_omitted_clock_reads_the_real_system_clock
    endpoint = APRV::VerifyReceiptEndpoint.new(
      trusted_roots: [TestSupport.fixture_certificate("receipt-root")],
      environment: APRV::Environment::SANDBOX
    )
    before = (Time.now.utc.to_r * 1000).to_i
    response = endpoint.verify_receipt_result(
      { "receipt-data" => [TestSupport.fixture_bytes("receipt")].pack("m0") }
    ).to_response
    after = (Time.now.utc.to_r * 1000).to_i
    rendered = response["receipt"]["request_date_ms"].to_i
    assert_operator rendered, :>=, before - 1000
    assert_operator rendered, :<=, after + 1000
  end

  # The seam must not be addable by accident: no receipt verdict depends on now.
  def test_receipt_verifier_rejects_a_clock_keyword
    assert_raises(ArgumentError) do
      APRV::ReceiptVerifier.new(
        trusted_roots: [TestSupport.fixture_certificate("receipt-root")],
        bundle_id: "com.example.app", clock: -> { Time.utc(2020, 1, 1) }
      )
    end
  end

  def test_verify_receipt_core_takes_no_clock
    assert_raises(ArgumentError) do
      APRV.verify_receipt_core(TestSupport.fixture_bytes("receipt"),
                               trusted_roots: [TestSupport.fixture_certificate("receipt-root")],
                               clock: -> { Time.utc(2020, 1, 1) })
    end
  end

  def test_a_clock_must_respond_to_call
    assert_raises(ArgumentError) do
      APRV::VerifyReceiptEndpoint.new(
        trusted_roots: [TestSupport.fixture_certificate("receipt-root")],
        environment: APRV::Environment::SANDBOX, clock: 42
      )
    end
  end

  def test_the_endpoint_clock_drives_only_the_request_date_triple
    receipt = TestSupport.fixture_bytes("receipt")
    roots = [TestSupport.fixture_certificate("receipt-root")]
    early = APRV::VerifyReceiptEndpoint.new(
      trusted_roots: roots, environment: APRV::Environment::SANDBOX,
      clock: -> { Time.utc(2019, 1, 1) }
    ).verify_receipt_result({ "receipt-data" => [receipt].pack("m0") }).to_response
    late = APRV::VerifyReceiptEndpoint.new(
      trusted_roots: roots, environment: APRV::Environment::SANDBOX,
      clock: -> { Time.utc(2099, 1, 1) }
    ).verify_receipt_result({ "receipt-data" => [receipt].pack("m0") }).to_response

    assert_equal 0, early["status"]
    assert_equal 0, late["status"]
    refute_equal early["receipt"]["request_date_ms"], late["receipt"]["request_date_ms"]
    assert_equal early["receipt"]["receipt_creation_date_ms"],
                 late["receipt"]["receipt_creation_date_ms"]
  end
end

File: node/test/clock.test.js
// Time: the JWS verifier judges no payload by its age, and the endpoint's
// optional `clock` seam moves request_date and nothing else. Both builds are
// exercised for the JWS side, because a rule that holds on only one of them
// is the drift web-parity.test.js is there to prevent.
import test from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import * as node from '../dist/index.js';
import * as web from '../dist/web/index.js';

const BUNDLE = 'com.example.app';

const fixture = (name) =>
  readFileSync(fileURLToPath(new URL(`../../fixtures/generated/${name}`, import.meta.url)));
const text = (name) => fixture(name).toString('ascii').trim();

// transaction.jws is signed at 2024-08-06T12:00:00Z; the assertions below
// are written against that instant, not against "now".
const SIGNED_AT = 1722945600000;
const TRANSACTION = text('transaction.jws');
const MINUTE = 60_000;

const BUILDS = [
  ['node', node],
  ['web', web],
];

function jwsVerifier(build, options) {
  return new build.JwsVerifier({
    trustedRoots: [fixture('jws-root.der')],
    bundleId: BUNDLE,
    acceptedEnvironments: ['Sandbox'],
    ...options,
  });
}

/** Awaits the web build's promises; the Node build returns values directly. */
async function outcome(run) {
  try {
    return { ok: await run() };
  } catch (error) {
    return { reason: error?.reason ?? `unexpected ${error}` };
  }
}

for (const [name, build] of BUILDS) {
  test(`${name}: a payload is never rejected for its age`, async () => {
    // Freshness is the caller's decision: a payload signed years ago still
    // verifies, and its signedDate is there for the caller to judge.
    const verdict = await outcome(() => jwsVerifier(build, {}).verifyTransaction(TRANSACTION));
    assert.equal(verdict.ok?.signedDate, SIGNED_AT);
  });

  test(`${name}: the removed maxSignedAgeMillis option is refused, not ignored`, () => {
    // A plain-JS caller still passing it must not lose the check silently.
    assert.throws(() => jwsVerifier(build, { maxSignedAgeMillis: MINUTE }), TypeError);
  });
}

// --- verifyReceipt endpoint ---------------------------------------------

test('endpoint: the clock drives request_date', () => {
  const fixedInstant = new Date('2025-03-04T05:06:07Z');
  const response = new node.VerifyReceiptEndpoint({
    trustedRoots: [fixture('receipt-root.der')],
    environment: 'Sandbox',
    clock: () => fixedInstant,
  })
    .verifyReceiptResult({ 'receipt-data': fixture('receipt.der').toString('base64') })
    .toResponse();
  assert.equal(response.status, 0);
  assert.equal(response.receipt.request_date_ms, String(fixedInstant.getTime()));
  assert.equal(response.receipt.request_date, '2025-03-04 05:06:07 Etc/GMT');
});

test('endpoint: omitting the clock stamps request_date from the system clock', () => {
  const before = Date.now();
  const response = new node.VerifyReceiptEndpoint({
    trustedRoots: [fixture('receipt-root.der')],
    environment: 'Sandbox',
  })
    .verifyReceiptResult({ 'receipt-data': fixture('receipt.der').toString('base64') })
    .toResponse();
  assert.equal(response.status, 0);
  const stamped = Number(response.receipt.request_date_ms);
  assert.ok(
    stamped >= before - 1000 && stamped <= Date.now() + 1000,
    `request_date_ms ${stamped} is not the current time`,
  );
});

test('endpoint: the clock does not move the receipt-chain verdict', () => {
  // Same fixed-window argument as the JWS case: validity is judged at the
  // receipt creation date, so the endpoint's answer is clock-independent.
  for (const clock of [
    undefined,
    () => new Date('2020-06-01T00:00:00Z'),
    () => new Date('2999-01-01T00:00:00Z'),
  ]) {
    const endpoint = (receipt) =>
      new node.VerifyReceiptEndpoint({
        trustedRoots: [fixture('receipt-expired-root.der')],
        environment: 'Sandbox',
        clock,
      })
        .verifyReceiptResult({ 'receipt-data': fixture(receipt).toString('base64') })
        .toResponse();
    assert.equal(endpoint('receipt-expired-historical.der').status, 0);
    assert.equal(endpoint('receipt-expired-fresh.der').status, 21003);
  }
});
